cloudsql/sqlserver/database-sql/cloudsql.go

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+func main() {
+
+	var err error
+
+	parsedTemplate, err = template.ParseFiles("templates/index.html")
+	if err != nil {
+		log.Fatalf("unable to parse template file: %s", err)
+	}
+
+	app := &App{}
+
+	// If the optional DB_TCP_HOST environment variable is set, it contains
+	// the IP address and port number of a TCP connection pool to be created,
+	// such as "127.0.0.1:1433". If DB_TCP_HOST is not set, a Unix socket
+	// connection pool will be created instead.
+	if os.Getenv("DB_TCP_HOST") != "" {
+		app.db, err = initTcpConnectionPool()
+		if err != nil {
+			log.Fatalf("initTcpConnectionPool: unable to connect: %s", err)
+		}
+	} else {
+		app.db, err = initSocketConnectionPool()
+		if err != nil {
+			log.Fatalf("initSocketConnectionPool: unable to connect: %s", err)
+		}
+	}
+
+	// Drop the votes table if it already exists.
+	if _, err = app.db.Exec(`DROP TABLE IF EXISTS votes;`); err != nil {
+		log.Fatalf("DB.Exec: unable to drop votes table: %s", err)
+	}
+	// Create the votes table.
+	if _, err = app.db.Exec(`CREATE TABLE votes
+	( vote_id int IDENTITY(1,1) PRIMARY KEY, time_cast DATETIME NOT NULL,
+	candidate CHAR(6) NOT NULL );`); err != nil {
+		log.Fatalf("DB.Exec: unable to create votes table: %s", err)
+	}
+
+	http.HandleFunc("/", app.indexHandler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+// recentVotes returns a slice of the last 5 votes cast.
+func recentVotes(app *App) ([]vote, error) {
+	var votes []vote
+	rows, err := app.db.Query(`SELECT TOP 5 candidate, time_cast FROM votes ORDER BY time_cast DESC`)
+	if err != nil {
+		return votes, fmt.Errorf("DB.Query: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		nextVote := vote{}
+		err := rows.Scan(&nextVote.Candidate, &nextVote.VoteTime)
+		if err != nil {
+			return votes, fmt.Errorf("Rows.Scan: %v", err)
+		}
+		votes = append(votes, nextVote)
+	}
+	return votes, nil
+}
+
+// currentTotals returns a templateData structure for populating the web page.
+func currentTotals(app *App) (templateData, error) {
+	// get total votes for each candidate
+	var tabVotes, spaceVotes uint
+	err := app.db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='TABS'`).Scan(&tabVotes)
+	if err != nil {
+		return templateData{}, fmt.Errorf("DB.QueryRow: %v", err)
+	}
+	err = app.db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='SPACES'`).Scan(&spaceVotes)
+	if err != nil {
+		return templateData{}, fmt.Errorf("DB.QueryRow: %v", err)
+	}
+
+	var voteDiffStr string = voteDiff(int(math.Abs(float64(tabVotes) - float64(spaceVotes)))).String()
+
+	latestVotesCast, err := recentVotes(app)
+	if err != nil {
+		return templateData{}, fmt.Errorf("recentVotes: %v", err)
+	}
+	return templateData{tabVotes, spaceVotes, voteDiffStr, latestVotesCast}, nil
```

Comment:
Generally a good practice to use named initialization for structs. That way, if the fields change at all, this declaration won't break. Would probably make sense to declare a separate variable at that point since it will be a bit longer.